apps/emqx_durable_storage/src/emqx_ds_lts.erl

```diff
@@ -0,0 +1,619 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_ds_lts).
+
+%% API:
+-export([
+    trie_create/1, trie_create/0, trie_restore/2, topic_key/3, match_topics/2, lookup_topic_key/2
+]).
+
+%% Debug:
+-export([trie_next/3, trie_insert/3, dump_to_dot/2]).
+
+-export_type([options/0, static_key/0, trie/0]).
+
+-include_lib("stdlib/include/ms_transform.hrl").
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-elvis([{elvis_style, variable_naming_convention, disable}]).
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+%% End Of Topic
+-define(EOT, []).
+-define(PLUS, '+').
+
+-type edge() :: binary() | ?EOT | ?PLUS.
+
+%% Fixed size binary
+-type static_key() :: non_neg_integer().
+
+-define(PREFIX, prefix).
+-type state() :: static_key() | ?PREFIX.
+
+-type varying() :: [binary() | ?PLUS].
+
+-type msg_storage_key() :: {static_key(), varying()}.
+
+-type threshold_fun() :: fun((non_neg_integer()) -> non_neg_integer()).
+
+-type persist_callback() :: fun((_Key, _Val) -> ok).
+
+-type options() ::
+    #{
+        persist_callback => persist_callback(),
+        static_key_size => pos_integer()
+    }.
+
+-record(trie, {
+    persist :: persist_callback(),
+    static_key_size :: pos_integer(),
+    trie :: ets:tid(),
+    stats :: ets:tid()
+}).
+
+-opaque trie() :: #trie{}.
+
+-record(trans, {
+    key :: {state(), edge()},
+    next :: state()
+}).
+
+%%================================================================================
+%% API funcions
+%%================================================================================
+
+%% @doc Create an empty trie
+-spec trie_create(options()) -> trie().
+trie_create(UserOpts) ->
+    Defaults = #{
+        persist_callback => fun(_, _) -> ok end,
+        static_key_size => 8
+    },
+    #{
+        persist_callback := Persist,
+        static_key_size := StaticKeySize
+    } = maps:merge(Defaults, UserOpts),
+    Trie = ets:new(trie, [{keypos, #trans.key}, set, public]),
+    Stats = ets:new(stats, [{keypos, 1}, set, public]),
```

Comment:
If I understood correctly, this eventually is called by the shard process (`emqx_ds_storage_layer`) when restoring the shard/opening a new generation, but that process' `terminate` callback doesn't seem to dump these tables.  Should the storage modules have an `on_close`/`on_terminate` callback for persisting data like this?

Reply:
The trie is restored by "replaying" the persisted records from the `trie` table, and the counters are updated during replay. So there's no need to persist the stats table. The records are written to the persistent storage as they are created, so there's nothing to be done in the terminate callback.